Client/WebAuthN/U2FExtensions.swift

```diff
@@ -979,20 +1186,78 @@ class U2FExtensions: NSObject {
             log.error(U2FErrorMessages.ErrorRegistration)
             return
         }
         fidoSignHandles.remove(at: index)
         fidoSignRequests.removeValue(forKey: handle)
         requestId.removeValue(forKey: handle)
-        YubiKitManager.shared.keySession.cancelCommands()
+        cleanup()
+    }
+    
+    private func executeKeyRequestWith(execution: @escaping () -> Void) {
+        if keyType == .accessory {
+            // Execute the request right away.
+            execution()
+        } else {
+            guard #available(iOS 13.0, *) else {
+                sendJSError()
+                return
+            }
+                        
+            guard let nfcSession = YubiKitManager.shared.nfcSession as? YKFNFCSession else {
+                sendJSError()
+                return
+            }
+            
+            self.nfcActive = true
+            
+            if nfcSession.iso7816SessionState == .open {
+                execution()
+                return
+            }
+            
+            // The ISO7816 session is started only when required since it's blocking the application UI with the NFC system action sheet.
+            YubiKitManager.shared.nfcSession.startIso7816Session()
+            
+            // Execute the request after the key(tag) is connected.
+            nfcSesionStateObservation = nfcSession.observe(\.iso7816SessionState, changeHandler: { [weak self] session, change in
+                if session.iso7816SessionState == .open {
+                    execution()
+                    self?.nfcActive = false
+                    self?.nfcSesionStateObservation = nil // remove the observation
+                }
+                
+                if self?.nfcActive ?? false, session.iso7816SessionState == .closed {
+                    // Session invalidated by user error is triggered for many cases
+                    // and is not a consistent way to track when user explicitly clicked
+                    // cancel. So we track for instances when nfcState is set to closed
+                    // without calling stopIso7816session
+                    self?.sendJSError()
+                    self?.nfcActive = false
+                    self?.nfcSesionStateObservation = nil // remove the observation
+                }
+            })
+        }
     }
     
     private func handleSessionStateChange() {
-        let sessionState = YubiKitManager.shared.keySession.sessionState
-        if sessionState == .open { // The key session is ready to be used.
+        defer {
+            observeSessionStateUpdates = true
+        }
+        
+        let accessoryState = YubiKitManager.shared.accessorySession.sessionState
+
+        if keyType == .nfc || accessoryState == .open { // The key session is ready to be used.
+            if accessoryState == .open {
+                if nfcActive, #available(iOS 13.0, *) {
+                    nfcActive = false
```

Comment:
Thinking whether we should `didSet` to `nfcActive` variable, so it calls `stopIso7816Session` always when it's set to false?
What do you think about it? Not 100% if we always want to stop the session when this property goes to `false`

Reply:
done!